Formula: add DSL to generate completions

max-ae opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR proposes an addition to the formula DSL to generate completions more easily.

The common pattern of 
```
(bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completion", "bash")
(zsh_completion/"_foo").write Utils.safe_popen_read(bin/"foo", "completion", "zsh")
(fish_completion/"foo.fish").write Utils.safe_popen_read(bin/"foo", "completion", "fish")
```
in homebrew/core could then be replaced by e.g.
```
generate_completions(shells: [:bash, :zsh])
```

Let me know if this is something you are interested in pursuing, then I will look into this further, including writing tests, docs, typecheck, and maybe also an AutoCorrector, although I have not much experience with RuboCop.

carlocab commented:
I haven't had the chance to take a close look at the implementation yet, but I think this might make for a nice addition to `brew`. Thanks for working on this!

MikeMcQuaid reviewed (commented):
This looks nice and makes sense to me so far! Do you have an idea of how many formulae would use this?

carlocab commented:
> Do you have an idea of how many formulae would use this?

I count 30, with
```bash
rg 'completion.*\.write.*completion.*sh' -l | wc -l
```

carlocab commented:
> The common pattern of
> 
> ```ruby
> (bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completion", "bash")
> (zsh_completion/"_foo").write Utils.safe_popen_read(bin/"foo", "completion", "zsh")
> (fish_completion/"foo.fish").write Utils.safe_popen_read(bin/"foo", "completion", "fish")
> ```

This seems to be the pattern for `go`-dependent formulae (presumably because they all use the same completions library). There's a slightly different pattern for at least some `rust`-dependent formulae (again, presumably because they all use the same library):

```ruby
(bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completions", "bash")
(zsh_completion/"_foo").write Utils.safe_popen_read(bin/"foo", "completions", "zsh")
(fish_completion/"foo.fish").write Utils.safe_popen_read(bin/"foo", "completions", "fish")
```

There's actually a bit more variation beyond that too. It would be nice if there was a way to handle most of the more common types generically (without making the actual handling here in `brew` too complicated).

max-ae commented:
Thanks for the feedback so far! I will look into this further.

> > Do you have an idea of how many formulae would use this?
> 
> I count 30, with
> 
> ```shell
> rg 'completion.*\.write.*completion.*sh' -l | wc -l
> ```

In my quick research, I counted 100+ formulae that could benefit from this. Keep in mind, some formulae also use something like this, so your `rg` expression does not capture all of them.
```
output = Utils.safe_popen_read(bin/"foo", "completion", "bash")
(bash_completion/"foo").write output
```

max-ae commented:
> There's actually a bit more variation beyond that too. It would be nice if there was a way to handle most of the more common types generically (without making the actual handling here in `brew` too complicated).

Do you have any suggestions on how to approach this?
My reasoning behind my draft implementation was to just go with "completion" as a default, and being able to override it with e.g. "completions" with a parameter, but I am open to other suggestions.

Trying possible commands or assuming them based on the language they are built with seems kind of... not nice in my opinion. What do you think?

max-ae commented:
I have added commits that include a typecheck signature as well as method documentation.

If someone could help or give a pointer on how to successfully implement a RuboCop, that would be greatly appreciated, given the complex logic (in my opinion, again: not much experience with RuboCop) needed to autocorrect the various completion generation patterns.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,37 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+  sig {
+    params(base_name: String, shells: T::Array[Symbol], binary: Pathname, cmd: String, shell_as_flag: T::Boolean).void
+  }
+  def generate_completions(base_name: name,
+                           shells: [:bash, :zsh, :fish],
+                           binary: bin/base_name,
+                           cmd: "completion",
+                           shell_as_flag: false)
```
Would be great to see how many formulae need each of these arguments?

MikeMcQuaid reviewed (commented):
This looks good so far! Would love a draft homebrew/core PR to use these APIs (which will fail CI, don't worry about it) so we can see what the usage will look like. Thanks!

Rylan12 commented:
> If someone could help or give a pointer on how to successfully implement a RuboCop, that would be greatly appreciated, given the complex logic (in my opinion, again: not much experience with RuboCop) needed to autocorrect the various completion generation patterns.

I think the best way to figure out the way RuboCop works is just to look at the existing examples in the [`rubocops/`](https://github.com/Homebrew/brew/tree/master/Library/Homebrew/rubocops) directory. I think the [`SafePopenCommands`](https://github.com/Homebrew/brew/blob/007059160f1a9d7afba296e9aa30ab52d4ef29b7/Library/Homebrew/rubocops/lines.rb#L240-L267) and [`OptionDeclarations`](https://github.com/Homebrew/brew/blob/007059160f1a9d7afba296e9aa30ab52d4ef29b7/Library/Homebrew/rubocops/lines.rb#L112-L193) checks may be a good starting point.

I think in terms of complexity, it looks like there are two main categories. First:

```ruby
(bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completions", "bash")
```

In this case, we should be able to use a `def_node_search` method (which I'll describe below in more detail). Like a Regex, we can use this to extract all the parts that could change (e.g. `foo`, `completions`, `bash`, etc). and then set up an autocorrector to automatically switch it to the new version.

Another case is:

```ruby
output = Utils.safe_popen_read(bin/"foo", "completion", "bash")
(bash_completion/"foo").write output
```

This is trickier to autocorrect since it's harder to track what's going on with `output`. What I'd recommend in this case is just detecting the `(bash_completion/"foo").write` part and showing an error without autocorrecting. The user will then need to make the change themselves which I think is fine. However, if it turns out that some formulae can't use the DSL and need to use this style, we may not be able to have a check for this.

---

In terms of methods that are available, take a look at what's in [`rubocops/shared/helper_functions.rb`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/rubocops/shared/helper_functions.rb). This has a lot of functions to help with finding certain methods and stuff like that. However, in this case, I think using [`def_node_search`](https://docs.rubocop.org/rubocop-ast/node_pattern.html#using-node-matcher-macros) will be super helpful. Basically, it defines a function that will search a node for all descendent nodes that match a certain pattern. You can use the `ruby-parse` shell command (you might need to install rubocop first) and the [Rubocop Node Pattern docs](https://docs.rubocop.org/rubocop-ast/node_pattern.html) to figure out what the pattern should be.

---

 Note also the rubocop will need to go in a separate PR because all of the homebrew/core instances need to be fixed before we merge the rubocop PR for CI to stay green. If you get stuck, feel free to open another PR and I'll be happy to help.

max-ae commented:
Thank you very much @Rylan12 for the extensive writeup! I'll start working on a RuboCop in a separate PR, then.

Following that, I will open an accompanying PR in homebrew/core incorporating the new API into formulae and to get a look at the usage.

SMillerDev commented:
This looks cool. Just wanted to note that there are a couple of formula using `Utils.safe_popen_read({ "SHELL" => "bash" }, bin/"helm", "completion", "bash")` too. So you might want to check if you can make those work here (maybe they don't need the env variables)

Rylan12 reviewed (commented):
Can `base_name` and `binary` be combined? I didn't see any instances where they were different in the example PR, and I feel like they should almost always be the same.

If there are cases when `binary` is different from `bin/base_name`, I'm guessing they're rare and so it might be fine to just require the old syntax. Thoughts?

Rylan12 commented:
> This looks cool. Just wanted to note that there are a couple of formula using `Utils.safe_popen_read({ "SHELL" => "bash" }, bin/"helm", "completion", "bash")` too. So you might want to check if you can make those work here (maybe they don't need the env variables)

What if we just always set the `SHELL` env variable in `safe_popen_read` to the corresponding shell whenever we're using `generate_completions`

chenrui333 commented:
It would be also good to add the options of generating the completions for individual shell (instead of generating for all shells).

chenrui333 commented:
> > The common pattern of
> > ```ruby
> > (bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completion", "bash")
> > (zsh_completion/"_foo").write Utils.safe_popen_read(bin/"foo", "completion", "zsh")
> > (fish_completion/"foo.fish").write Utils.safe_popen_read(bin/"foo", "completion", "fish")
> > ```
> 
> This seems to be the pattern for `go`-dependent formulae (presumably because they all use the same completions library). There's a slightly different pattern for at least some `rust`-dependent formulae (again, presumably because they all use the same library):
> 
> ```ruby
> (bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completions", "bash")
> (zsh_completion/"_foo").write Utils.safe_popen_read(bin/"foo", "completions", "zsh")
> (fish_completion/"foo.fish").write Utils.safe_popen_read(bin/"foo", "completions", "fish")
> ```
> 
> There's actually a bit more variation beyond that too. It would be nice if there was a way to handle most of the more common types generically (without making the actual handling here in `brew` too complicated).

Furthermore, they all use the same CLI framework (`spf13/cobra`) which has builtin completions support.

SMillerDev commented:
> It would be also good to add the options of generating the completions for individual shell (instead of generating for all shells).

Isn't that what the `shells` argument will allow?

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,45 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+  def generate_completions(base_name: name,
+                           shells: [:bash, :zsh, :fish],
+                           binary: bin/base_name,
+                           cmd: "completion",
```
```suggestion
                           subcmd: "completion",
```

or

```suggestion
                           parameter: "completion",
```

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,45 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+  }
+  def generate_completions(base_name: name,
+                           shells: [:bash, :zsh, :fish],
+                           binary: bin/base_name,
```
```suggestion
                           command: bin/base_name,
```

or

```suggestion
                           executable: bin/base_name,
```

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,45 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+    params(base_name: String, shells: T::Array[Symbol], binary: Pathname, cmd: String,
+           shell_prefix: T.nilable(T.any(Symbol, String))).void
+  }
+  def generate_completions(base_name: name,
```
```suggestion
  def generate_completions_from_command(base_name: name,
```

or 

```suggestion
  def generate_completions_from_executable(base_name: name,
```

to make it more obvious that it's going to be shelling out

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,45 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+                           shells: [:bash, :zsh, :fish],
+                           binary: bin/base_name,
+                           cmd: "completion",
+                           shell_prefix: nil)
```
```suggestion
                           shell_parameter: nil)
```

MikeMcQuaid reviewed (commented):
Some naming thoughts based on homebrew/core PR.

max-ae commented:
Thanks everyone for your thorough feedback! I will work on the suggested changes and subsequently update both the rubocop and example homebrew/core PR.

max-ae commented:
Hello everyone, I have updated the PR with some additional changes and applied the naming suggestions. Let me know what you think!
I have also updated the RuboCop PR and the homebrew/core example PR to reflect the changes made here.

carlocab commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,49 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+        "#{shell_parameter_format}#{shell}"
+      end
+
+      script_path.dirname.mkpath
```
I'm curious about why this is needed since many formulae seem to be able to do something like `script_path.write` directly.

max-ae replied:
I was wondering that too, but I am not all too familiar with all the `brew` internals.
Without this, something like this errors out, though.
```
❯ b reinstall k3sup -s
==> Cloning https://github.com/alexellis/k3sup.git
Updating /Users/Max/Library/Caches/Homebrew/k3sup--git
==> Checking out tag 0.12.0
HEAD is now at c59d67b Update Go crypo libraries and migrate to Go 1.18
HEAD is now at c59d67b Update Go crypo libraries and migrate to Go 1.18
==> Reinstalling k3sup
==> go build -ldflags=-s -w -X github.com/alexellis/k3sup/cmd.Version=0.12.0 -X github.com/alexellis/k3sup/cmd.GitCommit=c59d67b
Error: An exception occurred within a child process:
  Errno::ENOENT: No such file or directory @ rb_sysopen - /usr/local/Cellar/k3sup/0.12.0/etc/bash_completion.d/k3sup
```

carlocab reviewed (approved):
Thanks for your work, @max-ae!

carlocab commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,49 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+        "--shell=#{shell}"
+      elsif shell_parameter_format == :none
+        nil
+      else
+        "#{shell_parameter_format}#{shell}"
```
Which formulae did you have in mind for this case? I wonder if we should handle it at all. Or, if we will, let `shell_parameter_format` be a method (or a `Proc` object?) that takes an argument and spits out the right format.

max-ae replied:
For example, the `certigo` formula currently does this.
```ruby
# Install bash completion
output = Utils.safe_popen_read("#{bin}/certigo", "--completion-script-bash")
(bash_completion/"certigo").write output

# Install zsh completion
output = Utils.safe_popen_read("#{bin}/certigo", "--completion-script-zsh")
(zsh_completion/"_certigo").write output
```

This would turn into 
```ruby
generate_completions_from_executable(cmd: "", shells: [:bash, :zsh], shell_parameter_format: "--completion-script-")
``` 
and would not work without the current handling. 

Another example is the `earthly` formula.
```ruby
bash_output = Utils.safe_popen_read("#{bin}/earthly", "bootstrap", "--source", "bash")
(bash_completion/"earthly").write bash_output
zsh_output = Utils.safe_popen_read("#{bin}/earthly", "bootstrap", "--source", "zsh")
(zsh_completion/"_earthly").write zsh_output

generate_completions_from_executable(cmd: "bootstrap", shells: [:bash, :zsh], shell_parameter_format: "--source ")
# or 
generate_completions_from_executable(cmd: "bootstrap", shells: [:bash, :zsh], shell_parameter_format: "--source=")
```

carlocab replied:
In which case it might be useful to let `shell_parameter_format` be some sort of callable so we can accommodate formulae like `rabbitmq`:
```ruby
    (bash_completion/"rabbitmqadmin.bash").write Utils.safe_popen_read("#{sbin}/rabbitmqadmin", "--bash-completion")
```
Maybe `yield` will do enough here. But feel free not to handle this if it's too complicated.

Rylan12 replied:
Another option might be to use a placeholder like `@@SHELL@@` (maybe that's too long) so we can do (in addition:

```ruby
    generate_completions_from_executable(base_name: "rabbitmqadmin", shell_parameter: "--@@SHELL@@-completion")
```

If we use a block, I think it might be smart to have the block set the list of arguments, rather than just the shell parameter. It feels more flexible, that way. Something like this:

```ruby
    generate_completions_from_executable do |shell|
      [bin/"rabbitmqadmit", "--#{shell}-completion"]
    end
```

max-ae replied:
I like your suggestions, but I feel like this extension to cover even more formulae with the new syntax might warrant a followup PR.

Either someone else can take this up, or I might in the future when I familiarized myself a bit more with blocks/`Proc`s 😄

Rylan12 commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,49 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+  #
+  # @param base_name [String] the base name of the generated completion script. Defaults to the formula name.
+  # @param shells [Array<Symbol>] the shells to generate completion scripts for. Defaults to `[:bash, :zsh, :fish]`.
+  # @param binary [Pathname] the binary to use for generating the completion scripts. Defaults to the binary with the
```
The names here need to be updated

Rylan12 commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,49 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+                                           shells: [:bash, :zsh, :fish],
+                                           executable: bin/base_name,
+                                           cmd: "completion",
+                                           shell_parameter_format: nil)
```
Thoughts on `shell_parameter_prefix` as a name? That way the string case discussed above is clearer. Also, it's clearer what the difference between `nil` and `:none` is

max-ae replied:
`_prefix` was the name I originally went with, but it does not quite fit, especially with `:none`, as that will not pass anything to the command (as opposed to the default, `nil`, that will simply pass the shell name). 
The `:none` case is needed for the formulae that infer the shell to generate for from the environment variable SHELL.

But I am open to other naming suggestions for this argument!

Rylan12 replied:
Yeah, fair point. I guess I'm okay with this or the above suggestions since I can't think of something better right now

Rylan12 reviewed (commented):
This looks really good!

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,50 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+  # executable with the name of the formula.
+  # @param cmd [String] the command to pass to the `binary`. Defaults to 'completion'.
+  # @param shell_parameter_format [String]/[Symbol] specify how `shells` should each be passed
+  # to the `executable`. Takes either a String representing a prefix, or one of [:flag, :arg, :none].
```
Would be good to have examples of each one and what they translate to re: calling commands.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,50 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+  #
+  # @param base_name [String] the base name of the generated completion script. Defaults to the formula name.
+  # @param shells [Array<Symbol>] the shells to generate completion scripts for. Defaults to `[:bash, :zsh, :fish]`.
+  # @param executable [Pathname] the executable to use for generating the completion scripts. Defaults to the
```
Kinda feel like this should be mandatory, we tend to not call executables implicitly without naming them as it makes the DSL less readable.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1625,6 +1625,50 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+  # @param shells [Array<Symbol>] the shells to generate completion scripts for. Defaults to `[:bash, :zsh, :fish]`.
+  # @param executable [Pathname] the executable to use for generating the completion scripts. Defaults to the
+  # executable with the name of the formula.
+  # @param cmd [String] the command to pass to the `binary`. Defaults to 'completion'.
```
```suggestion
  # @param subcmd [String] the subcommand to pass to the `binary`. Defaults to 'completion'.
```

max-ae commented:
I have made the requested changes of making `executable` and `subcmd` mandatory arguments. 

I have also added a test, although it fails as-is. Anyone have any idea what I am missing here? 😄

Rylan12 commented on Library/Homebrew/test/formula_spec.rb:
```diff
@@ -1996,4 +1996,28 @@ def install
+
+    it "generates completion scripts" do
+      f.brew { f.install }
+      expect(f.bash_completion/"foo").to be_a_file
+      expect(f.zsh_completion/"_foo").to be_a_file
+      expect(f.fish_completion/"foo.fish").to be_a_file
```
This test formula's name is `testball`, so you'll need to update these checks to look for the correct file:

```suggestion
      expect(f.bash_completion/"testball").to be_a_file
      expect(f.zsh_completion/"_ testball").to be_a_file
      expect(f.fish_completion/"testball.fish").to be_a_file
```

Or, change the `generate_completions_from_executable` call to specify the base name:

```ruby
generate_completions_from_executable(bin/"foo", "test", base_name: "foo")
```

max-ae replied:
Thanks 🤦🏻‍♂️

Rylan12 reviewed (approved):
Great work, @max-ae! This looks fantastic!

max-ae commented on Library/Homebrew/formula.rb:
```diff
@@ -1677,11 +1677,11 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+    params(executable: Pathname, subcmd: T.nilable(T.any(String, T::Array[String])), base_name: String, shells: T::Array[Symbol],
            shell_parameter_format: T.nilable(T.any(Symbol, String))).void
   }
   def generate_completions_from_executable(executable,
```
It does! I had something like this in mind too, and just tested it. I am not sure about the type sigil, though...

max-ae commented:
Dammit! Now I know again why I wrote it like this the first time :D Done.

max-ae commented on Library/Homebrew/formula.rb:
```diff
@@ -1622,6 +1622,98 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
+      popen_read_args.flatten!
+
+      script_path.dirname.mkpath
+      script_path.write Utils.safe_popen_read({ "SHELL" => shell.to_s }, *popen_read_args)
```
I tried both of your suggestions before, unfortunately, `typecheck` will error out in both cases...

Rylan12 replied:
Huh, I've never seen that error before and we do this everywhere 😅. Not worth fixing, so let's stick with it as is.
